addons/sale_loyalty/tests/test_program_rules.py

```diff
@@ -231,36 +228,77 @@ def test_program_rules_validity_dates_and_uses(self):
+        msg = "The promo shouldn't have been applied as it is expired."
+        self.assertEqual(len(order.order_line.ids), 2, msg)
 
+        self.immediate_promotion_program.write({'date_to': future_day})
+        self._auto_rewards(order, self.immediate_promotion_program)
+        msg = "The promo should have been applied we're between the validity dates."
+        self.assertEqual(len(order.order_line.ids), 3, msg)
+
+        # Test date_from (no date_to)
         self.immediate_promotion_program.write({
-            'date_to': date.today() + timedelta(days=2),
+            'date_from': future_day, 'date_to': False,
         })
-        order = self.env['sale.order'].create({'partner_id': self.steve.id})
+        self._auto_rewards(order, self.immediate_promotion_program)
+        msg = "The promo shouldn't have been applied as it is not active yet."
+        self.assertEqual(len(order.order_line.ids), 2, msg)
+
+        self.immediate_promotion_program.write({'date_from': past_day})
+        self._auto_rewards(order, self.immediate_promotion_program)
+        msg = "The promo should have been applied we're between the validity dates."
+        self.assertEqual(len(order.order_line.ids), 3, msg)
+
+        # Test date_from and date_to
+        self.immediate_promotion_program.write({'date_from': past_day, 'date_to': future_day})
+        self._auto_rewards(order, self.immediate_promotion_program)
+        msg = "The promo should have been applied as we're between the validity dates"
+        self.assertEqual(len(order.order_line.ids), 3, msg)
+
+        self.immediate_promotion_program.write({
+            'date_from': today + timedelta(days=1),
+            'date_to': future_day,
+        })
+        self._auto_rewards(order, self.immediate_promotion_program)
+        msg = "The promo offer shouldn't have been applied as it is not active yet."
+        self.assertEqual(len(order.order_line.ids), 2, msg)
+
+        self.immediate_promotion_program.write({
+            'date_from': past_day,
+            'date_to': today - timedelta(days=1),
+        })
+        self._auto_rewards(order, self.immediate_promotion_program)
+        msg = "The promo offer shouldn't have been applied as it is expired."
+        self.assertEqual(len(order.order_line.ids), 2, msg)
+
+        self.immediate_promotion_program.write({'date_from': today, 'date_to': today})
+        self._auto_rewards(order, self.immediate_promotion_program)
+        msg = "The promo should have been applied as today is a valid starting and ending date."
+        self.assertEqual(len(order.order_line.ids), 3, msg)
+
+    def test_program_rules_number_of_uses(self):
+        # Test case: Based on the number of allowed uses
+        self.immediate_promotion_program.write({
+            'limit_usage': True,
+            'max_usage': 1,
+        })
+        order = self.empty_order
         order.write({'order_line': [
             (0, False, {
```

Comment:
```suggestion
            Command.create({
```
You can replace the old magic commands by the new "cleaner" API